### PR TITLE
planner: use ordered index with is null predicate | tidb-test=pr/2368

### LINKED ISSUE #54188: optimizer does not consider using ordered index scan when `is null` is used
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
CREATE TABLE `testnull` (
       `object_id` bigint(20) unsigned NOT NULL,
       `a` bigint(20) DEFAULT NULL,
       `b` bigint(20) DEFAULT NULL,
       PRIMARY KEY (`object_id`) /*T![clustered_index] CLUSTERED */,
       KEY `ab` (`a`,`b`)
);

explain select * from testnull where a is null order by b;
```

### 2. What did you expect to see? (Required)
execution plan use index ab and keep order on b to avoid sorting

### 3. What did you see instead (Required)
```sql
mysql> explain select * from testnull where a is null order by b;
+--------------------------+---------+-----------+--------------------------------+---------------------------------------------------+
| id                       | estRows | task      | access object                  | operator info                                     |
+--------------------------+---------+-----------+--------------------------------+---------------------------------------------------+
| Sort_5                   | 10.00   | root      |                                | test.testnull.b                                   |
| └─IndexReader_9          | 10.00   | root      |                                | index:IndexRangeScan_8                            |
|   └─IndexRangeScan_8     | 10.00   | cop[tikv] | table:testnull, index:ab(a, b) | range:[NULL,NULL], keep order:false, stats:pseudo |
+--------------------------+---------+-----------+--------------------------------+---------------------------------------------------+
3 rows in set (0.01 sec)

mysql> explain select /*+ order_index(testnull, ab) */ * from testnull where a is null order by b;
ERROR 1815 (HY000): Internal : Can't find a proper physical plan for this query
```

### 4. What is your TiDB version? (Required)

```sql
mysql> select @@version\g;
+--------------------+
| @@version          |
+--------------------+
| 8.0.11-TiDB-v8.1.0 |
+--------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by ari-e**
### What problem does this PR solve?

Issue Number: close #54188 

Problem Summary: Properly classify columns with null predicates (e.g. `WHERE a IS NULL`) as constant so index selection can take advantage of that to satisfy a sort in https://github.com/pingcap/tidb/blob/432bb79f9732cb89a0be220ca93415df37a4849e/pkg/planner/core/find_best_task.go#L791.

### What changed and how does it work?
Added checks for is null predicate during planning and fill corresponding data structures marking that column as a constant.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/casetest/index/index_test.go [hunk 1]
```diff
@@ -148,3 +148,15 @@ func TestRowFunctionMatchTheIndexRangeScan(t *testing.T) {
 		tk.MustQuery(tt).Sort().Check(testkit.Rows(output[i].Result...))
```
Comment (hawkingrei) [inline pkg/planner/core/casetest/index/index_test.go:148]: Please run the ```make bazel_prepare``` and update the changement.

Comment (michaelmdeng) [reply pkg/planner/core/casetest/index/index_test.go:148]: @hawkingrei thanks! unfortunately @ari-e is out for a bit and I don't have permissions to update this PR, so I've addressed the comment separately in #54290

**Comment by mzhang77** [comment]
Please ignore this PR and review https://github.com/pingcap/tidb/pull/54290 instead.

**Comment by ari-e** [comment]
Picked this PR back up now that I'm back from vacation. So please consider this one and we'll abandon #54290 #54512.

I've fixed the unit test failures and bazel build issues. Now the 3 issues left are failed integration tests. I suspect though that these are failing because the plan output is legitimately different with my change. The outputs of those 3 tests are:

```
run test [util/ranger] err: sql:explain format='brief' select * from t where a = 1 and (b is null or b = 2) and c > 1;: failed to run query
"explain format='brief' select * from t where a = 1 and (b is null or b = 2) and c > 1;"
 around line 176,
we need(316):
explain format='brief' select * from t where a = 1 and (b is null or b = 2) and c > 1;
id      estRows task    access object   operator info
IndexReader     0.07    root            index:Selection
└─Selection     0.07    cop[tikv]               gt(util__ranger.t.c, 1)
  └─IndexRangeScan      0.20    cop[tikv]       table:t, index:a(a, b, c)       range:[1 NULL,1 NULL], [1
but got(316):
explain format='brief' select * from t where a = 1 and (b is null or b = 2) and c > 1;
id      estRows task    access object   operator info
IndexReader     0.67    root            index:IndexRangeScan
└─IndexRangeScan        0.67    cop[tikv]       table:t, index:a(a, b, c)       range:(1 NULL 1,1 NULL +inf], (1 2 1,1 2 +inf], keep order:false, stats:pseudo
```

```
run test [util/ranger] err: sql:explain format='brief' select * from t where a = 1 and (b is null or b = 2) and c > 1;: failed to run query
"explain format='brief' select * from t where a = 1 and (b is null or b = 2) and c > 1;"
 around line 176,
we need(316):
explain format='brief' select * from t where a = 1 and (b is null or b = 2) and c > 1;
id      estRows task    access object   operator info
IndexReader     0.07    root            index:Selection
└─Selection     0.07    cop[tikv]               gt(util__ranger.t.c, 1)
  └─IndexRangeScan      0.20    cop[tikv]       table:t, index:a(a, b, c)       range:[1 NULL,1 NULL], [1
but got(316):
explain format='brief' select * from t where a = 1 and (b is null or b = 2) and c > 1;
id      estRows task    access object   operator info
IndexReader     0.67    root            index:IndexRangeScan
└─IndexRangeScan        0.67    cop[tikv]       table:t, index:a(a, b, c)       range:(1 NULL 1,1 NULL +inf], (1 2 1,1 2 +inf], keep order:false, stats:pseudo
```

```
run test [func_group] err: sql:explain format="brief"
select max(a3) from t1 where a2 is null;: failed to run query
"explain format="brief"
select max(a3) from t1 where a2 is null;"
 around line 330,
we need(421):
explain format="brief"
select max(a3) from t1 where a2 is null;
id      estRows task    access object   operator info
StreamAgg       1.00    root            funcs:max(func_group.t1.a3)->Column#7
└─TopN  1.00    root            func_group.t1.a3:desc, offset:0, count:1
  └─IndexReader 1.00    root            index:TopN
    └─TopN      1.00    cop[tikv]               func_group.t1.a3:desc, offset:0, count:1
      └─IndexRangeScan  2.00    cop[tikv]       table:t1, index:k1(a2, a3)      range:[N
but got(421):
explain format="brief"
select max(a3) from t1 where a2 is null;
id      estRows task    access object   operator info
StreamAgg       1.00    root            funcs:max(func_group.t1.a3)->Column#7
└─Limit 1.00    root            offset:0, count:1
  └─IndexReader 1.00    root            index:Limit
    └─Limit     1.00    cop[tikv]               offset:0, count:1
      └─IndexRangeScan  1.00    cop[tikv]       table:t1, index:k1(a2, a3)      range:[NULL -inf,NULL +inf], keep order:true, desc
```

@hawkingrei can you provide any guidance for whether these seem like legitimate cases where the integration test needs to be updated, or whether my PR needs to change to not affect these query plans?

**Review by hawkingrei — APPROVED** [review]

**Comment by hawkingrei** [comment]
@ari-e LGTM

But you should update the result in the ```tests/integrationtest```.

for example
```
cd tests/integrationtest
./run-tests.sh -r executor/merge_join
```

**Comment by ari-e** [comment]
@hawkingrei I fixed the integration test in `tests/integrationtest` but it looks like there's some test external to this repo that is being run that is still failing called `func_group` which is part of `ghpr_mysql_test`. Do you know where that is defined? [Jenkins log](https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fghpr_mysql_test/detail/ghpr_mysql_test/13156/pipeline)

File: pkg/util/ranger/detacher.go [hunk 2]
```diff
@@ -714,8 +731,12 @@ func ExtractEqAndInCondition(sctx *rangerctx.RangerContext, conditions []express
 		if ma == nil {
 			if accesses[i] != nil {
 				if allEqOrIn(accesses[i]) {
-					newConditions = append(newConditions, accesses[i])
 					columnValues[i] = extractValueInfo(accesses[i])
+					if columnValues[i] != nil && columnValues[i].value != nil && columnValues[i].value.IsNull() {
+						accesses[i] = nil
+					} else {
+						newConditions = append(newConditions, accesses[i])
+					}
```
Comment (winoros) [inline pkg/util/ranger/detacher.go:739]: With these code blocks, we could add some test cases like `where a is null or a in (1, 2, 3)` for `TestConvertRangeToPoint` in `tests/integrationtest/t/planner/core/casetest/integration.test`

The `points2EqOrInCond` called in line#737 should also be updated?

Comment (ari-e) [reply pkg/util/ranger/detacher.go:739]: Hi @winoros thanks for the comments.

> With these code blocks, we could add some test cases like `where a is null or a in (1, 2, 3)` for `TestConvertRangeToPoint` in `tests/integrationtest/t/planner/core/casetest/integration.test`

Sure

> The `points2EqOrInCond` called in line#737 should also be updated?

I'm pretty sure it's a no-op to do this since intersecting a null point with non-null points works as expected, but I'll do it anyways to short-circuit the logic.

**Comment by hawkingrei** [comment]
@ari-e Please sync with master.

**Comment by ari-e** [comment]
Rebased on master @hawkingrei @winoros

**Comment by winoros** [comment]
The `idc-jenkins-ci-tidb/mysql-test` passed. I think your pr should have passed all tests now. Other failures are not caused by your main code changes.

**Review by winoros — APPROVED** [review]
